modules/rostests/apitests/advapi32/Hash.c

```diff
@@ -0,0 +1,298 @@
+                        g_TestString.Length) == g_TestString.Length,
+       "A_SHA_CTX::Buffer incorrect after A_SHAUpdate\n");
+    ok(RtlCompareMemory(ctx.State,
+                        g_ctxSHA1StateInit,
+                        RTL_FIELD_SIZE(A_SHA_CTX, State) == RTL_FIELD_SIZE(A_SHA_CTX, State)),
+       "A_SHA_CTX::State incorrect after A_SHAUpdate\n");
+    ok(ctx.Count[0] == 0 && ctx.Count[1] == g_TestString.Length,
+       "A_SHA_CTX::Count incorrect after A_SHAUpdate\n");
+
+    /* Test A_SHAFinal */
+    pfnA_SHAFinal(&ctx, Result);
+    ok(RtlCompareMemoryUlong(ctx.Buffer, sizeof(ctx.Buffer), 0) == sizeof(ctx.Buffer),
+       "A_SHA_CTX::Buffer incorrect after A_SHAFinal\n");
+    ok(RtlCompareMemory(ctx.State,
+                        g_ctxSHA1StateInit,
+                        RTL_FIELD_SIZE(A_SHA_CTX, State)) == RTL_FIELD_SIZE(A_SHA_CTX, State),
+       "A_SHA_CTX::State incorrect after A_SHAFinal\n");
+    ok(ctx.Count[0] == 0 && ctx.Count[1] == 0, "A_SHA_CTX::Count incorrect after A_SHAFinal\n");
+    ok(RtlCompareMemory(Result, g_aucSHA1Result, A_SHA_DIGEST_LEN) == A_SHA_DIGEST_LEN,
+       "SHA1 result incorrect\n");
+}
+
+static void Test_MD5()
+{
+    FN_MD5Init* pfnMD5Init;
+    FN_MD5Update* pfnMD5Update;
+    FN_MD5Final* pfnMD5Final;
+    DECLSPEC_ALIGN(4) MD5_CTX ctx;
+
+    /* Load functions */
+    pfnMD5Init = (FN_MD5Init*)GetProcAddress(g_hAdvapi32, "MD5Init");
+    pfnMD5Update = (FN_MD5Update*)GetProcAddress(g_hAdvapi32, "MD5Update");
+    pfnMD5Final = (FN_MD5Final*)GetProcAddress(g_hAdvapi32, "MD5Final");
+
+    if (pfnMD5Init == NULL || pfnMD5Update == NULL || pfnMD5Final == NULL)
+    {
+        skip("advapi32.dll!MD5XXX not found\n");
+        return;
+    }
+    
+    /* Test MD5Init */
+    pfnMD5Init(&ctx);
+    ok(ctx.Count[0] == 0 && ctx.Count[1] == 0, "MD5_CTX::Count incorrect after MD5Init\n");
+    ok(RtlCompareMemory(ctx.State,
+                        g_aulMD5Or4StateInit,
+                        RTL_FIELD_SIZE(MD5_CTX, State)) == RTL_FIELD_SIZE(MD5_CTX, State),
+       "MD5_CTX::State incorrect after MD5Init\n");
+
+    /* Test MD5Update */
+    pfnMD5Update(&ctx, (PUCHAR)g_TestString.Buffer, g_TestString.Length);
+    ok(RtlCompareMemory(ctx.Buffer,
+                        g_TestString.Buffer,
+                        g_TestString.Length) == g_TestString.Length,
+       "MD5_CTX::Buffer incorrect after MD5Update\n");
+    ok(RtlCompareMemory(ctx.State,
+                        g_aulMD5Or4StateInit,
+                        RTL_FIELD_SIZE(MD5_CTX, State) == RTL_FIELD_SIZE(MD5_CTX, State)),
+       "MD5_CTX::State incorrect after MD5Update\n");
+    ok(ctx.Count[0] == g_TestString.Length * CHAR_BIT && ctx.Count[1] == 0,
+       "MD5_CTX::Count incorrect after MD5Update\n");
+
+    /* Test MD5Final */
+    pfnMD5Final(&ctx);
+    ok(ctx.Count[0] == 0x200 && ctx.Count[1] == 0,
+       "MD5_CTX::Count [0x%08lX, 0x%08lX] is incorrect after MD5Final, expect [0x200, 0]\n",
```

Comment:
> `Count` field still incorrect after `MD(4/5)Final`

```
g_TestString.Length == 28
g_TestString.Length * CHAR_BIT == 224 == 0xE0

0x200 == 512
```

ReactOS' `MD5Final()` does not modify `Count`, hence the unchanged value.

Note that MD4 and MD5 use 512-bit blocks.
Could you test what happens with some other input lengths? Does Windows always set `0x200` (or a multiple)?

Reply:
Maybe I will handle that (correct the `Count` output and add test for larger content to hash) in the future, but not planned on this PR.